workspaces/adventure-pack/goodies/typescript/Iterator.prototype.reduce/index.ts

```diff
@@ -0,0 +1,39 @@
+import "../Iterator.prototype.toIterable";
+import { iteratorPrototype } from "../Iterator.prototype";
+
+declare global {
+  interface Iterator<T> {
+    reduce<U>(
+      this: Iterator<T>,
+      callbackFn: (accumulator: U, value: T, index: number) => U,
+      initialValue?: U,
+    ): U;
+  }
+}
+
+iteratorPrototype.reduce ??= function <T extends U, U>(
+  this: Iterator<T>,
+  callbackFn: (accumulator: U, value: T, index: number) => T,
+  initialValue?: U,
+): U {
+  let index = 0;
+  let accumulator: U;
+
+  if (initialValue === undefined) {
```

Comment:
I'm not sure what the spec says exactly, but by checking against undefined like this we don't distinguish between `iterator.reduce(someCallback)` and `iterator.reduce(someCallback, undefined)`. If we wanted to distinguish between these two cases we could check the length of the arguments object: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/arguments/length

Reply:
ack!